clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -422,7 +434,17 @@ public boolean hasUnsent() {
                                         break;
                                     } else {
                                         RecordBatch batch = deque.pollFirst();
-                                        batch.close();
+                                        if (transactionState != null) {
+                                            log.debug("Dest: {} : pid: {}, epoch: {}, Assigning sequence for {}-{} : {}", node,
+                                                    transactionState.pidAndEpoch().pid, transactionState.pidAndEpoch().epoch,
+                                                    batch.topicPartition.topic(), batch.topicPartition.partition(),
+                                                    transactionState.sequenceNumber(batch.topicPartition));
+                                            batch.closeWithTransactionState(transactionState.pidAndEpoch().pid,
+                                                    transactionState.pidAndEpoch().epoch,
+                                                    transactionState.sequenceNumber(batch.topicPartition));
+                                        } else {
+                                            batch.close();
```

Comment:
`close()` now only get called in `abortBatches`. I'm wondering if we could collapse these two with `close(TransactionState)` in which we will update the state if it is not null?